File: SignalScribe.Vue/src/lib/detectedMode.ts
/** Mirrors SignalScribe.Enums.DetectedMode. */
export enum DetectedMode {
  Unknown = 0,
  AnalogFm = 1,
  Afsk1200 = 2,
  Fsk9600 = 3,
  Pocsag = 4,
  Flex = 5,
  Dmr = 6,
  DStar = 7,
  Ysf = 8,
  P25Phase1 = 9,
  Nxdn = 10,
  M17 = 11,
  DigitalUnknown = 12,
}

/** Short label for the chip, the longer read on hover, and how loudly to say it. */
export const DETECTED_MODES: Record<DetectedMode, { label: string; detail: string; color: string }> = {
  [DetectedMode.Unknown]: {
    label: "unknown",
    detail: "Not enough signal to tell what modulation this was.",
    color: "surface-variant",
  },
  [DetectedMode.AnalogFm]: {
    label: "FM",
    detail: "Ordinary analog narrowband FM — the usual thing on 2m.",
    color: "surface-variant",
  },
  [DetectedMode.Afsk1200]: {
    label: "APRS",
    detail: "1200 baud packet. Confirmed by decoding an AX.25 frame, not guessed.",
    color: "info",
  },
  [DetectedMode.Fsk9600]: {
    label: "9600 packet",
    detail: "9600 baud G3RUH packet data.",
    color: "info",
  },
  [DetectedMode.Pocsag]: {
    label: "POCSAG",
    detail: "Pager traffic. Two-level FSK with a wide ±4.5 kHz shift.",
    color: "info",
  },
  [DetectedMode.Flex]: {
    label: "FLEX",
    detail: "Motorola FLEX paging.",
    color: "info",
  },
  [DetectedMode.Dmr]: {
    label: "DMR",
    detail: "Four-level C4FM on DMR's ±1944/±648 Hz plan, two slots of TDMA.",
    color: "purple",
  },
  [DetectedMode.DStar]: {
    label: "D-STAR",
    detail: "GMSK at 4800 bps. The header carries callsigns in plain text.",
    color: "purple",
  },
  [DetectedMode.Ysf]: {
    label: "Fusion",
    detail: "Yaesu System Fusion, C4FM.",
    color: "purple",
  },
  [DetectedMode.P25Phase1]: {
    label: "P25",
    detail: "P25 Phase 1, C4FM on the ±1800/±600 Hz plan.",
    color: "purple",
  },
  [DetectedMode.Nxdn]: {
    label: "NXDN",
    detail: "NXDN four-level FSK.",
    color: "purple",
  },
  [DetectedMode.M17]: {
    label: "M17",
    detail: "M17 — the open digital voice mode, C4FM with Codec 2.",
    color: "purple",
  },
  [DetectedMode.DigitalUnknown]: {
    label: "digital",
    detail:
      "Discrete symbol levels, so something is sending data — but not a mode we can name yet. " +
      "Several modes share a deviation plan, so naming one from the levels alone would be a guess.",
    color: "warning",
  },
};

export function modeLabel(mode: DetectedMode): string {
  return DETECTED_MODES[mode]?.label ?? "unknown";
}

export function modeDetail(mode: DetectedMode): string {
  return DETECTED_MODES[mode]?.detail ?? "";
}

export function modeColor(mode: DetectedMode): string {
  return DETECTED_MODES[mode]?.color ?? "surface-variant";
}

/** True when the mode names the frequency specifically — mirrors DetectedModeExtensions.IsIdentified. */
export function isIdentified(mode: DetectedMode): boolean {
  return (
    mode !== DetectedMode.Unknown &&
    mode !== DetectedMode.AnalogFm &&
    mode !== DetectedMode.DigitalUnknown
  );
}

/** Modes worth putting a chip on. Analog FM is the default and would be noise on every row. */
export function isWorthShowing(mode: DetectedMode): boolean {
  return mode !== DetectedMode.Unknown && mode !== DetectedMode.AnalogFm;
}

/** Everything the <ModeChip> component needs to render one tag. */
export interface ModeChipSpec {
  label: string;
  detail: string;
  color: string;
  icon: string;
  variant: "tonal" | "outlined";
}

function plain(mode: DetectedMode): ModeChipSpec {
  return {
    label: modeLabel(mode),
    detail: modeDetail(mode),
    color: modeColor(mode),
    icon: isIdentified(mode) ? "mdi-waveform" : "mdi-help-rhombus-outline",
    variant: "tonal",
  };
}

/**
 * Modulation measured on one transmission, and whether it is what this channel normally carries.
 * A mode that differs from the channel's usual one is the interesting case, exactly as with tones:
 * a different system sharing the frequency, or someone set to the wrong mode.
 */
export function transmissionModeChip(mode: DetectedMode, channelMode: DetectedMode | null): ModeChipSpec | null {
  if (!isWorthShowing(mode)) return null;

  if (channelMode != null && channelMode !== mode && isIdentified(mode) && isIdentified(channelMode)) {
    return {
      label: modeLabel(mode),
      color: "warning",
      icon: "mdi-swap-horizontal",
      variant: "tonal",
      detail: `${modeDetail(mode)} This channel normally carries ${modeLabel(channelMode)}, so this is either a different system on the same frequency or someone set to the wrong mode.`,
    };
  }

  return plain(mode);
}

/**
 * What a channel carries. An operator-set modulation wins over the measured one and shows solid;
 * a measured mode shows outlined with an ear, the same convention the tone column already uses.
 */
export function channelModeChip(
  declared: DetectedMode | null,
  measured: DetectedMode | null,
): ModeChipSpec | null {
  if (declared != null) {
    return {
      label: modeLabel(declared),
      color: modeColor(declared),
      detail: `Set by you as ${modeLabel(declared)}. ${modeDetail(declared)}`,
      icon: "mdi-waveform",
      variant: "tonal",
    };
  }

  if (measured == null || !isWorthShowing(measured)) return null;

  return {
    label: modeLabel(measured),
    color: modeColor(measured),
    detail: `Measured from the air. ${modeDetail(measured)}`,
    icon: "mdi-ear-hearing",
    variant: "outlined",
  };
}

File: SignalScribe/Enums/DetectedMode.cs
namespace SignalScribe.Enums;

/// <summary>
/// What a transmission's modulation was measured to be, capture-side, from the discriminator.
///
/// This is a *measurement*, not a decode: naming the mode is what lets a channel be labelled and a
/// data frequency be understood rather than merely discarded. The definitive identification for the
/// digital voice modes comes later, from sync-pattern correlation in the per-mode framers — the
/// classifier's job is cheap triage, and <see cref="DigitalUnknown"/> is a legitimate answer.
/// </summary>
public enum DetectedMode
{
    /// <summary>No confident verdict. Too little signal, or nothing matched well enough to commit.</summary>
    Unknown = 0,

    /// <summary>Ordinary analog NBFM — continuous deviation, no symbol structure. The default on 2m.</summary>
    AnalogFm = 1,

    /// <summary>Bell 202 AFSK, 1200 baud: APRS and classic packet.</summary>
    Afsk1200 = 2,

    /// <summary>G3RUH 9600 baud direct FSK packet.</summary>
    Fsk9600 = 3,

    /// <summary>POCSAG paging, 2FSK at ±4.5 kHz.</summary>
    Pocsag = 4,

    /// <summary>FLEX paging, 2FSK/4FSK.</summary>
    Flex = 5,

    /// <summary>DMR Tier II: 4FSK, 4800 sym/s, ±1944/±648 Hz, 2-slot TDMA.</summary>
    Dmr = 6,

    /// <summary>D-STAR: GMSK, 4800 bps.</summary>
    DStar = 7,

    /// <summary>System Fusion: C4FM, 4800 sym/s.</summary>
    Ysf = 8,

    /// <summary>P25 Phase 1: C4FM, 4800 sym/s, ±1800/±600 Hz.</summary>
    P25Phase1 = 9,

    /// <summary>NXDN: 4FSK, 2400 (NXDN48) or 4800 (NXDN96) sym/s.</summary>
    Nxdn = 10,

    /// <summary>M17: C4FM with Codec 2 — the one digital voice mode that is open end to end.</summary>
    M17 = 11,

    /// <summary>
    /// Clearly not analog — discrete symbol levels were measured — but which mode it is could not be
    /// settled. Deliberately a first-class answer: guessing a specific mode from level structure alone
    /// is the same mistake as rounding a hum onto the nearest CTCSS tone.
    /// </summary>
    DigitalUnknown = 12,
}

/// <summary>Pure predicates over <see cref="DetectedMode"/>, kept separate from the capture code so they can be tested alone.</summary>
public static class DetectedModeExtensions
{
    /// <summary>
    /// Whether the mode is named specifically enough to say what a frequency *is*. This is what may
    /// rescue a transmission from the voice gate and bring a channel into existence, so it excludes
    /// all three ways of not knowing:
    /// <list type="bullet">
    /// <item><see cref="DetectedMode.AnalogFm"/> — an unmodulated carrier and a burst of noise both
    /// land here, and treating that as recognition would let every kerchunk and spur create a
    /// channel, which is exactly what the voice gate exists to prevent.</item>
    /// <item><see cref="DetectedMode.DigitalUnknown"/> — knowing a signal is digital is not knowing
    /// what it is. An APRS burst reads this way until the soft-TNC decodes a frame, and creating a
    /// channel on it would put 144.390 straight back to recording every packet forever, which is the
    /// failure <see cref="SignalScribe.Analysis.ChannelVoiceAudit"/> was written to undo.</item>
    /// <item><see cref="DetectedMode.Unknown"/> — no verdict at all.</item>
    /// </list>
    /// </summary>
    public static bool IsIdentified(this DetectedMode mode) =>
        mode is not (DetectedMode.Unknown or DetectedMode.AnalogFm or DetectedMode.DigitalUnknown);

    /// <summary>
    /// Whether the mode carries data rather than speech. Data channels are labelled and decoded to
    /// text but do not retain clips — the payload is the text, not the sound.
    /// </summary>
    public static bool IsData(this DetectedMode mode) => mode is
        DetectedMode.Afsk1200 or DetectedMode.Fsk9600 or DetectedMode.Pocsag or DetectedMode.Flex;

    /// <summary>Whether the mode is digital *voice* — framed, and vocodable into an ordinary clip.</summary>
    public static bool IsDigitalVoice(this DetectedMode mode) => mode is
        DetectedMode.Dmr or DetectedMode.DStar or DetectedMode.Ysf
        or DetectedMode.P25Phase1 or DetectedMode.Nxdn or DetectedMode.M17;
}
